Counter: STM32 Counter RTC Driver

erwango opened the pull request:
Provide STM32 driver for Counter RTC API.
Additionally provides basic sample to exercise the API.

IP (hence driver) doesn't support wrap related functions.
Tested on disco_l475_iot1, nucleo_f401re, stm32f3_disco, nucleo_f746zg

Fixes #11373

erwango committed:
drivers/counter: stm32: Provide STM32 RTC driver using counter API

Provides a STM32 RTC driver using new counter API.
Driver does not support wrap related functions (set, get, ..)

Fixes #11373

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/counter: stm32 rtc: Use dts to get RTC clock information

RTC clock information should be provided by dts files

Fixes #10451

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

ydamigos commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,342 @@
+
+	/* Return number of seconds since RTC init */
+	ts -= T_TIME_OFFSET;
+	printk("ts: %lld\n", ts);
```
Please use LOG_DBG.

ydamigos commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,342 @@
+		ticks += now;
+	}
+
+	printk("Set Alarm: %d\n", ticks);
```
Please use LOG_DBG.

ydamigos commented on samples/drivers/counter_rtc/src/main.c:
```diff
@@ -0,0 +1,55 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
```
Please move the file under samples/drivers commit 814d929dbc22749a0f19cf98bccdf51c57805296.

ydamigos commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,342 @@
+		return -EIO;
+	}
+
+	LL_RTC_DisableWriteProtection(RTC);
```
We could remove this.

ydamigos commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,342 @@
+	LL_RTC_ALMA_Disable(RTC);
+
+	if (LL_RTC_ALMA_Init(RTC, LL_RTC_FORMAT_BIN, &rtc_alarm) != SUCCESS) {
+		return -EIO;
```
Should we call `LL_RTC_EnableWriteProtection(RTC)` before return the error?

ydamigos commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,342 @@
+	}
+
+#ifdef RTC_CR_BYPSHAD
+	LL_RTC_EnableShadowRegBypass(RTC);
```
`LL_RTC_DisableWriteProtection` function should be called before.

ydamigos commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,342 @@
+		}
+
+		LL_RTC_ClearFlag_ALRA(RTC);
+		LL_RTC_DisableIT_ALRA(RTC);
```
`LL_RTC_DisableWriteProtection` function should be called before.

dbkinder reviewed (dismissed):
+1 for doc change

erwango committed:
samples/drivers: Provides basic sample for counter alarm

Provide basic sample to exercise counter alarm API.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Adding modifications required to get counter tests passed.
Note that tests are failed when MPU is on. Opening a ticket for that matter

erwango commented:
Shippable complains of following errors:
- .set_alarm = aon_counter_qmsi_set_alarm (waits https://github.com/zephyrproject-rtos/zephyr/pull/9276 I guess)
-  File "/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/gen_cfb_font_header.py", line 10, in <module>
    from PIL import ImageFont
ImportError: No module named 'PIL'.
Need to dig into this one

erwango commented:
recheck

erwango commented:
recheck

erwango commented:
recheck

MaureenHelm commented:
@erwango, please rebase and remove all commits that aren't yours (the commits you need from @nordic-krch 	should already be in the topic branch). Sanity check also should pass now.

erwango commented:
@MaureenHelm, rebased. I've removed @nordic-krch's commits but kept yours since they are needed to enable test on STM32 based boards, is that ok?
@ydamigos, can you review again?

MaureenHelm commented on samples/drivers/counter_rtc/sample.yaml:
```diff
@@ -0,0 +1,13 @@
+sample:
+  name: Counter RTC Driver Sample
```
Why are you calling this a "counter_rtc" sample? Why not just "counter"?

Also need to add license/copyright header to new files and a sample README.rst.

erwango replied:
Sure, so I finally changed to counter/alarm as it specifically plays with alarm.
README added

MaureenHelm reviewed (changes requested):
> @MaureenHelm, rebased. I've removed @nordic-krch's commits but kept yours since they are needed to enable test on STM32 based boards, is that ok?

Yes

@nordic-krch 	Are you ok with the test changes?

ydamigos reviewed (dismissed):
LGTM

erwango commented:
@MaureenHelm, @nordic-krch, rebased on top of the branch

nordic-krch commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -117,7 +126,11 @@ void test_wrap_alarm_instance(const char *dev_name)
 
 void test_wrap_alarm(void)
 {
+#ifdef CONFIG_COUNTER_MCUX_RTC
```
it would be better to handle that in the test. If first call to `set_wrap` (now called `set_top_value`) returns `-ENOTSUP` test can stop the counter and leave without failure. We should avoid device specific defines in api test.

nordic-krch commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -157,19 +170,15 @@ void test_single_shot_alarm_instance(const char *dev_name)
 	err = counter_disable_ch_alarm(dev, &alarm_cfg);
 	zassert_equal(0, err, "Counter disabling alarm failed\n");
 
-	err = counter_set_wrap(dev, counter_get_max_wrap(dev), NULL, NULL);
```
you shouldn't remove that but rather accept -ENOTSUP error. It resets counter to default state. Once this is removed, on some platforms following tests may fail.

nordic-krch replied:
maybe it would be good to split this test to part which does not require `set_wrap` (not `set_top_value`) and one that requires. Because now, you kind of limiting this test for platforms that support `set_wrap`.

nordic-krch commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -295,7 +308,11 @@ void test_all_channels_instance(const char *str)
 void test_all_channels(void)
 {
+#ifdef CONFIG_COUNTER_MCUX_RTC
+	ztest_test_skip();
```
why you've disabled this test? If device supports only one channel that should work as well.

carlescufi commented:
@erwango can you please rebase this on top of the new topic-counters now that the NXP PR is merged?

galak commented:
> @erwango can you please rebase this on top of the new topic-counters now that the NXP PR is merged?

Done

nordic-krch commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,347 @@
+	return ticks;
+}
+
+static int rtc_stm32_set_alarm(struct device *dev,
```
api needs to be aligned: 
```
static inline int counter_set_channel_alarm(struct device *dev, u8_t chan_id,
				      const struct counter_alarm_cfg *alarm_cfg)
```

nordic-krch commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,347 @@
+}
+
+
+static int rtc_stm32_disable_alarm(struct device *dev,
```
api needs to be aligned

erwango replied:
done

nordic-krch commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,347 @@
+}
+
+
+static u32_t rtc_stm32_get_wrap(struct device *dev)
```
renamed to `get_top_value`

erwango replied:
done

nordic-krch commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,347 @@
+}
+
+
+static int rtc_stm32_set_wrap(struct device *dev, u32_t ticks,
```
`set_top_value`

erwango replied:
done

nordic-krch commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,347 @@
+		LL_RTC_EnableWriteProtection(RTC);
+
+		if (alarm_cfg && alarm_cfg->handler) {
+			alarm_cfg->handler(dev, alarm_cfg, now);
```
handler syntax changed.

galak replied:
can you explain that further?

erwango replied:
done

nordic-krch commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,352 @@
+
+	DEV_DATA(dev)->alarm_cfg = alarm_cfg;
+
+	if (chan_id >= DEV_CFG(dev)->counter_info.channels) {
```
this check is redundant, its already performed in the wrapper.

nordic-krch commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,352 @@
+	u32_t now = rtc_stm32_read(dev);
+	u32_t ticks = alarm_cfg->ticks;
+
+	DEV_DATA(dev)->alarm_cfg = alarm_cfg;
```
api changed so that `alarm_cfg` is not persistent anymore. You need to store user data and handler.

erwango committed:
boards: nucleo_f746zg: Configure and document counter RTC

Configure RTC on nucleo_f746zg.
It is supported through counter RTC API.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers: counter: bump cmake minimum version

counter_basic_api test is cmake outdated.
Set it to newest version.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

nordic-krch commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,344 @@
+void rtc_stm32_isr(void *arg)
+{
+	struct device *const dev = (struct device *)arg;
+	const struct counter_alarm_cfg *alarm_cfg = DEV_DATA(dev)->alarm_cfg;
```
i don't see any place where `alarm_cfg` pointer would be set.

erwango replied:
right, good catch txs

erwango commented:
Please do not merge yet, as I haven't been able to test latest version

erwango commented:
recheck

erwango commented:
@nordic-krch , CI is failed due to failing qemu_x86/tests/net/socket/select/ which is unrelated to my changes (as I could test).

nordic-krch commented on samples/drivers/counter/alarm/src/main.c:
```diff
@@ -0,0 +1,66 @@
+	printk("!!! Alarm !!!\n");
+	printk("Now: %d\n", now_sec);
+
+	counter_disable_channel_alarm(counter_dev, ALARM_CHANNEL_ID);
```
you shouldn't need to call that as alarms are single shot (i will prepare PR soon to clarify that in counter.h)

nordic-krch commented on samples/drivers/counter/alarm/src/main.c:
```diff
@@ -0,0 +1,66 @@
+
+	counter_disable_channel_alarm(counter_dev, ALARM_CHANNEL_ID);
+
+	config->ticks = config->ticks +	counter_us_to_ticks(counter_dev, DELAY);
```
ticks are relative as set in line 47 thus you don't need to do that. If you want to keep precise timing you should set `absolute` to true and keep that line.

erwango replied:
This was actually the purpose of the sample, cf README:
> At each alarm expiry, a new alarm is configured with a delay increased by 2 seconds.

Of course, we can remove an set always the same duration, but I wanted to play with alarm duration. Let me double the delay at each expiry to make it easier to get.

nordic-krch replied:
ok, i see. i missed that.

nordic-krch commented on samples/drivers/counter/alarm/src/main.c:
```diff
@@ -0,0 +1,66 @@
+	}
+
+	while (1) {
+	}
```
please add `k_sleep(K_FOREVER)`. In case logger is used this loop will block execution of it (since it runs in lowest priority thread).

nordic-krch commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -0,0 +1,354 @@
+		LL_RTC_EnableWriteProtection(RTC);
+
+		if (data && data->callback) {
+			data->callback(dev, 0, now,
```
clear `data->callback` before this call to free channel.

nordic-krch replied:
i still don't see `data->callback` being cleared. If you call `counter_set_channel_alarm` from that callback you'll get `-EBUSY` error.

galak reviewed (changes requested):
Should we add the STM counter to tests/drivers/counter/counter_basic_api/test_counter.c device list?

galak commented on boards/arm/stm32l476g_disco/stm32l476g_disco.yaml:
```diff
@@ -10,3 +10,4 @@ ram: 96
 flash: 1024
 supported:
   - gpio
+  - rtc
```
should this also have counter ? (same for other board yaml)

erwango replied:
I think this is the limit of having one API for multiple purposes.
Do we list IPs or API here?

erwango replied:
ok, will add

erwango committed:
tests: counter: Update to support STM32 RTC counter

Update counter_basic_api test to support STM32 RTC counter.
Add device to list and skipp not supported test cases.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
soc: stm32: Convert CONFIG_RTC leftover to DT_

Convert CONFIG_RTC_0_IRQ_PRI to DT_ namespace.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
recheck

erwango commented:
Trying to solve timeout issue, removing last commit (fixing CONFIG_ to DT_ for IRQ_PRIO).

nashif commented:
> Trying to solve timeout issue, removing last commit (fixing CONFIG_ to DT_ for IRQ_PRIO).

the timeout is because you are changing too many boards (yml file) and that causes them to build with --all

erwango commented:
> the timeout is because you are changing too many boards (yml file) and that causes them to build with --all

Ok, splitting into several PRs to make it more digest to shippable.

galak commented:
Closing as we merged a split version of this.
